test/cql-pytest/README.md

```diff
@@ -42,3 +42,163 @@ Additional useful pytest options, especially useful for debugging tests:
+   diverge from Cassandra's implementation - or just explaining the
+   difference in text.
+
+6. **Think about risky cases, don't just randomize**: One of the benefits
+   of a developer also developing the tests during development of a feature
+   (instead of someone else doing it later) is that specific edge cases
+   can be considered and tested. For example, consider some operation taking
+   a string. If a developer knows that an empty string or a very long
+   string required special code and are at risk of being mis-handled or
+   are at risk to break during some future refactoring, the developer should
+   write separate tests for these cases.
+   If, instead, we write a test that loops 1000 times testing a random
+   string of length 10 - the result will be slow **and** will also miss
+   the interesting cases - neither the empty or very long strings will
+   come up as random draws of strings of length 10.
+   Another danger of randomized tests is that they tend to obscure what
+   is actually being tested: For example, a reviewer of the test may think
+   the empty string is included in the test, while actually it isn't.
+   Randomized "fuzz" testing has its benefits, but it is almost always the
+   wrong thing to do in the context of the cql-pytest framework. We should
+   probably have a separate framework (or at least separate files) for
+   these tests.
+
+7. **Write tests, not a test library**: Developers are often told that
+   long functions are evil, and are tempted to take maybe-useful sections
+   of code from their test function and split them out to utility functions.
+   Having a small number of these utility functions is indeed convenient,
+   and we have some in `util.py`, `nodetool.py` and `rest_api.py`. However
+   please resist the urge to add more and more of these utility functions.
+   Utility functions are bad for several reasons. First, they make it
+   harder to read tests - any reader will understand what "cql.execute(...)"
+   does, but not be familiar with dozens of obscure utilty functions.
+   Second, when a utility function is written for the benefit of a single
+   test, it is often much less general than its author thought, and when
+   it comes time to reuse it in a second test, it turns out it needs to be
+   duplicated or changed, and the result is many confusingly-similar
+   utility functions, or utility functions with many confusing parameters.
+   We've seen this happening in other frameworks such as dtest.
+   If you believe something _should_ be a utility function, start by
+   putting it inside the single test file that needs it - and only move
+   it to util.py if several test files can benefit from it.
+   At the time of this writing, cql-pytest has over 20,000 lines of test
+   code, and around of 500 lines of library code. Please keep this ratio.
+   We're writing a collection of tests - not a library.
+
+8. **Do not over-design**: Continuing the "we're writing tests, not a
```

Comment:
Again. Arguable. It's convenient to make easier to write tests (many/most developers don't want to in the first place). Also same rules of code duplication apply to test code as in non-test code. But this is just my opinion.